lib/Dialect/Calyx/Transforms/CompileControl.cpp

```diff
@@ -0,0 +1,190 @@
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/OperationSupport.h"
+
+using namespace circt;
+using namespace calyx;
+using namespace mlir;
+
+/// A helper function to create constants in the HW dialect.
+static hw::ConstantOp createConstant(OpBuilder &builder, Location loc,
+                                     size_t bitWidth, size_t value) {
+  return builder.create<hw::ConstantOp>(
+      loc, APInt(bitWidth, value, /*unsigned=*/true));
+}
+
+/// Given some number of states, returns the necessary bit width
+/// TODO(Calyx): Probably a better built-in operation?
+static size_t getNecessaryBitWidth(size_t numStates) {
+  APInt apNumStates(64, numStates);
+  size_t log2 = apNumStates.ceilLogBase2();
+  return log2 > 1 ? log2 : 1;
+}
+
+/// Creates a RegisterOp, with input and output port bit widths defined by
+/// `width`.
+static RegisterOp createRegister(OpBuilder &builder, ComponentOp &component,
+                                 size_t width, StringRef name) {
+  auto *context = builder.getContext();
+  builder.setInsertionPointToStart(component.getBody());
+  return builder.create<RegisterOp>(component->getLoc(),
+                                    StringAttr::get(context, name), width);
+}
+
+/// Generates a latency-insensitive FSM to realize a sequential operation. This
+/// is done by initializing GroupGoOp values for the enabled groups in the
+/// SeqOp, and then creating a new Seq GroupOp with the given FSM. Each step in
+/// the FSM is guarded by the done operation of the group currently being
+/// executed. After the group is complete, the FSM is incremented. This SeqOp is
+/// then replaced in the control with an Enable statement referring to the new
+/// Seq GroupOp.
+static void visitSeqOp(SeqOp &seq, ComponentOp &component) {
+  auto wires = component.getWiresOp();
+  Block *wiresBody = wires.getBody();
+
+  auto &seqOps = seq.getBody()->getOperations();
+  assert(llvm::all_of(seqOps, [](auto &&op) { return isa<EnableOp>(op); }) &&
+         "The SeqOp should only contain EnableOps in this pass.");
+  // This should be the number of enable statements + 1 since this is the
+  // maximum value the FSM register will reach.
+  size_t fsmBitWidth = getNecessaryBitWidth(seqOps.size() + 1);
+
+  OpBuilder builder(component->getRegion(0));
+  auto fsmRegister = createRegister(builder, component, fsmBitWidth, "fsm");
+  auto fsmIn = fsmRegister.getResult(0);
+  auto fsmWriteEn = fsmRegister.getResult(1);
+  auto fsmOut = fsmRegister.getResult(4);
+
+  builder.setInsertionPointToStart(wiresBody);
+  auto oneConstant = createConstant(builder, wires->getLoc(), 1, 1);
+
+  // Create the new compilation group to replace this SeqOp.
+  builder.setInsertionPointToEnd(wiresBody);
+  auto seqGroup =
+      builder.create<GroupOp>(wires->getLoc(), builder.getStringAttr("seq"));
+  Block *seqGroupBody = new Block();
+  seqGroup->getRegion(0).push_back(seqGroupBody);
+
+  // Guarantees a unique SymbolName for the group.
```

Comment:
Neat! But shouldn't this happen when the `GroupOp` gets created? Maybe I'm not understanding...

Reply:
The goal here is to ensure that if we have compile two `seq` ops, we'll produce two unique names, e.g. `seq`, `_seq`. I was looking through the documentation, and I saw that we can guarantee this with the symbol table insertion, rather than writing up my own name mangler. Is there a better way to do this that I'm missing?

Reply:
I was under the impression any time an op with the Symbol trait is created within an op with the SymbolTable trait, the name uniquing would occur, but I may be wrong. I was just looking at the code and it looks like the trait may just raise an error rather than automatically uniquing the name: https://github.com/llvm/llvm-project/blob/64c2b2c69d61dbb6459037a7bfddf29e1f280c8f/mlir/lib/IR/SymbolTable.cpp#L417

Reply:
Yeah that's how I came to [this](https://mlir.llvm.org/doxygen/SymbolTable_8cpp_source.html#l00157), and saw a similar use case [here](https://mlir.llvm.org/doxygen/AsyncParallelFor_8cpp_source.html#l00380).